pkg/ioutils/readers.go

```diff
@@ -170,3 +170,22 @@ func subsequentCloseWarn(name string) {
 		log.G(context.TODO()).Errorf("stack trace: %s", string(debug.Stack()))
 	}
 }
+
+type readerCtx struct {
+	ctx context.Context
+	r   io.Reader
+}
+
+func NewCtxReader(ctx context.Context, r io.Reader) io.Reader {
```

Comment:
Need a doc string here.

Also I think something like an `io.PipeReader` would be a better for the use case which you can call `CloseWithError(ctx.Err())` on and handle cancellation a little more completely (rather than be potentially blocked on reading the underlying reader).

e.g.

```go
pr, pw := io.Pipe()

go func() {
  <-ctx.Done()
  pr.CloseWithError(ctx.Err())
}()

// ...
```

Or we can forgo the pipe and just close the reader normally when the context is cancelled and check after the `io.Copy` if it was cancelled.

Reply:
I'm not sure if the pipe actually solves anything though.

If the underlying reader `Read` method blocks indefinitely, then there's actually no way of stopping it, other than  calling `Close` on the reader if it's actually also an `io.Closer` and hoping that ends all pending `Read`s.

I would prefer to avoid doing any `Close` inside the wrapper, because of the `defer closer.Close()` being actually the most common way of handling calling close. Doing that could potentially result in `Close` being called twice (first in the wrapper `Read` call, and then in the defer), which is implementation specific and can lead to subtle bugs like https://github.com/moby/moby/pull/46418.

Reply:
Added a `CopyCtx` as a replacement for the `io.Copy` as a best-effort recovery  in a situation where `Read` blocks completely - in this case the calling goroutine won't be blocked, only the child goroutine that does the `io.Copy`.